### ISSUE TITLE
Backend looses its workers and stops running jobs after a while

### ISSUE THREAD

**Issue by JeroenvIS** [netdisco/netdisco#1236]
## Expected Behavior

We expect jobs like "discoverall", "macwalk" and "arpwalk" to be run according to the configured schedules.

## Current Behavior

After starting the backend, jobs are being run fine for quite a while (at least several days). After some unknown time, jobs are not being run any more - we can see that the timestamps for last_discover, last_macsuck and last_arpnip are not updated for any device any more.

## Possible Solution

Not sure. We have been running for a long time (months? years?) with a cron job that just restarts the netdisco-backend every night, but that's not a real solution.

## Steps to Reproduce (for bugs)

We have not been able to actively reproduce. If we don't restart the backend on a daily basis, it will take somewhere between a couple of days to a month at maximum before the backend fails to run jobs.

When jobs aren't run any more, we don't see anything in the logs/backend.log pointing to something that went wrong.

It appears that the "nd2 master" process is dead and no worker processes are active any more, when the issues starts; a `ps -ef | grep netdisco` output looks like this:

```
netdisco@linux786:~/logs$ ps -ef | grep netdisco
netdisco 2362887       1  0 Jul04 ?        00:19:30 netdisco-backend
netdisco 2362888 2362887  0 Jul04 ?        00:23:14 [nd2: master] <defunct>

```
For comparison, the same output on our test server with nightly restarts of the backend looks like this:
```
netdisco 1152781       1  0 03:00 ?        00:02:40 netdisco-backend
netdisco 1152782 1152781  0 03:00 ?        00:00:21 nd2: master
netdisco 1152843 1152782  0 03:00 ?        00:00:00 nd2: #1 sched: idle
netdisco 1152844 1152782  0 03:00 ?        00:02:13 nd2: #2 mgr: idle
netdisco 1274291 1152782  0 10:38 ?        00:00:00 nd2: #5 poll: #45413336: arpnip x.y.28.8
netdisco 1274294 1152782  3 10:38 ?        00:00:24 nd2: #10 poll: #45413318: arpnip a.b.0.203
netdisco 1277134 1152782 26 10:50 ?        00:00:23 nd2: #3 poll: #45413320: arpnip a.b.0.205
netdisco 1277241 1152782 25 10:50 ?        00:00:20 nd2: #6 poll: #45413319: arpnip a.b.0.204
netdisco 1277245 1152782 23 10:50 ?        00:00:15 nd2: #4 poll: #45413321: arpnip a.b.0.206
netdisco 1277287 1152782  0 10:51 ?        00:00:00 nd2: #9 poll: idle
netdisco 1277288 1152782  0 10:51 ?        00:00:00 nd2: #7 poll: idle
netdisco 1277296 1152782  1 10:51 ?        00:00:00 nd2: #8 poll: idle
```

## Context

We're using Netdisco to keep an accurate inventory. Lots of scripts in our network depend on Netdisco for an up-to-date inventory, eg for which devices to pull stats, backup configs, check and modify port settings, etc. Having stale data in Netdisco makes all the other scripts less reliable.

## Your Environment

* Netdisco version used: 2.070001
* SNMP::Info version used: 3.95

## Config info (deployment.yml)

Most relevant parts:
```
# daemon will keep netdisco up to date on this schedule
# `````````````````````````````````````````````````````
schedule:
  discoverall:
    when: '5 * * * *'
  macwalk:
    when:
      min: 20
  arpwalk:
    when:
      min: 50
  nbtwalk:
    when: '0 8,13,21 * * *'
  expire:
    when: '20 23 * * *'

# number of SNMP pollers to run in parallel (in netdisco-daemon)
# ``````````````````````````````````````````````````````````````
workers:
  pollers: 20

# amount parallel DNS resolution for node names
# `````````````````````````````````````````````
dns:
  max_outstanding: 50
```

## Device information
Not related to any specific device

**Comment by JeroenvIS** [comment]
For what it's worth, we're checking median age for several jobs across devices now (and expect all to be under one hour when Netdisco is doing its job(s)) - it was a nice SQL exercise. This is current output with a defunct nd2 master process since just over 3 days:

```
netdisco@linux786:~$ netdisco-do psql -e "select percentile_cont(0.5) within group (order by age(now(), last_discover)) as median_last_discover_age, percentile_cont(0.5) within group (order by age(now(), last_macsuck)) as median_last_macsuck_age, percentile_cont(0.5) within group (order by age(now(), last_arpnip)) as median_last_arpnip_age from device"
[2559335] 2024-07-15 09:00:31  info App::Netdisco version 2.070001 loaded.
[2559335] 2024-07-15 09:00:31  info psql:  started at Mon Jul 15 11:00:31 2024
 median_last_discover_age | median_last_macsuck_age | median_last_arpnip_age
--------------------------+-------------------------+------------------------
 3 days 12:43:25.734403   | 3 days 12:38:23.899978  | 3 days 12:10:07.284626
(1 row)

[2559335] 2024-07-15 09:00:31  info psql: finished at Mon Jul 15 11:00:31 2024
[2559335] 2024-07-15 09:00:31  info psql: status done: psql session closed.
```

**Comment by ollyg** [comment]
Hi @JeroenvIS thanks for the ticket! 

What's interesting to me, is that this is the nd2-master process, which is `netdisco-backend-fg` where the actual job queue lives (and not the other child processes which put and pick jobs on/off the queue, or workers doing jobs). So I wonder if it's a bug in the queueing library or the way we're using it. Being `<defuct>` would indicate it crashed.

Would it be at all possible to run `netdisco-backend-fg` instead of `netdisco-backend` for a few days, and watch it crash? So using something like screen/tmux and tee to grab any dying gasp output. I appreciate this might be difficult as you use it operationally.

I will also have a look to see if we can get `netdisco-backend` to restart `netdisco-backend-fg` either periodically (yuck) or by detecting the failure itself (slightly less but still yuck).

**Comment by JeroenvIS** [comment]
On both my servers, the crash hasn't happened yet. We're monitoring them closely, for active process counts as well as median age for the last_{discover,macsuck,arpnip} age - so I expect that we'll notice quickly when one of them dies again.

After the first crash, I'll see if I can run the `netdisco-backend-fg` in the foreground. I always have a screen session running on these servers.

**Comment by JeroenvIS** [comment]
Interesting new data point: our test server still has the normal number of nd2 processes, but the median age of last_* timestamps is increasing - no new jobs are being run.

A `ps -ef | grep nd2` output that I collected yesterday afternoon, when I first noticed that the last_* ages were increasing:

```
root@linux470:~# ps -ef | grep nd2
netdisco    1557    1556  0 Jul17 ?        00:12:02 nd2: master
netdisco    2058    1557  0 Jul17 ?        00:00:19 nd2: #1 sched: idle
netdisco    2059    1557  0 Jul17 ?        01:16:21 nd2: #2 mgr: idle
netdisco  706531    1557  0 02:07 ?        00:00:08 nd2: #4 poll: #46018965: discover 172.31.146.75
netdisco  721856    1557  0 03:05 ?        00:00:08 nd2: #5 poll: #46020613: discover 172.31.146.75
netdisco  742810    1557  0 04:26 ?        00:00:08 nd2: #6 poll: #46022261: discover 172.31.146.75
netdisco  766692    1557  0 06:03 ?        00:00:08 nd2: #10 poll: #46023827: discover 172.31.146.75
netdisco  770596    1557  0 06:16 ?        00:00:08 nd2: #7 poll: #46025133: discover 172.31.146.75
netdisco  809747    1557  0 08:55 ?        00:00:08 nd2: #9 poll: #46025940: discover 172.31.146.75
root      922467 3318307  0 16:48 pts/7    00:00:00 grep nd2
netdisco 2016712    1557  0 Jul22 ?        00:00:06 nd2: #8 poll: #45704906: discover 172.31.146.75
netdisco 3646658    1557  0 Jul26 ?        00:00:06 nd2: #3 poll: #45880669: discover 172.31.146.76
```
and the same command just now:
```
root@linux470:~# ps -ef | grep nd2
netdisco    1557    1556  0 Jul17 ?        00:12:10 nd2: master
netdisco    2058    1557  0 Jul17 ?        00:00:20 nd2: #1 sched: idle
netdisco    2059    1557  0 Jul17 ?        01:16:42 nd2: #2 mgr: idle
netdisco  706531    1557  0 Jul30 ?        00:00:08 nd2: #4 poll: #46018965: discover 172.31.146.75
netdisco  721856    1557  0 Jul30 ?        00:00:08 nd2: #5 poll: #46020613: discover 172.31.146.75
netdisco  742810    1557  0 Jul30 ?        00:00:08 nd2: #6 poll: #46022261: discover 172.31.146.75
netdisco  766692    1557  0 Jul30 ?        00:00:08 nd2: #10 poll: #46023827: discover 172.31.146.75
netdisco  770596    1557  0 Jul30 ?        00:00:08 nd2: #7 poll: #46025133: discover 172.31.146.75
netdisco  809747    1557  0 Jul30 ?        00:00:08 nd2: #9 poll: #46025940: discover 172.31.146.75
root     1260410 3318307  0 16:35 pts/7    00:00:00 grep nd2
netdisco 2016712    1557  0 Jul22 ?        00:00:06 nd2: #8 poll: #45704906: discover 172.31.146.75
netdisco 3646658    1557  0 Jul26 ?        00:00:06 nd2: #3 poll: #45880669: discover 172.31.146.76
```

So somehow several discover processes are polling the same device and this started to build up slowly. I'd expect them not to run in parallel in the first place, but when that does happen, could it perhaps result in a deadlock?

@ollyg any insights from your end, with this new data?

**Comment by JeroenvIS** [comment]
Looks like a locking issue indeed:

```
netdisco@linux470:~$ netdisco-do psql
[1264283] 2024-07-31 14:52:33  info App::Netdisco version 2.070001 loaded.
[1264283] 2024-07-31 14:52:33  info psql:  started at Wed Jul 31 16:52:33 2024
psql (12.19 (Ubuntu 12.19-0ubuntu0.20.04.1), server 15.7 (Ubuntu 15.7-1.pgdg22.04+1))
WARNING: psql major version 12, server major version 15.
         Some psql features might not work.
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
Type "help" for help.

netdisco_linux470=> select pid,
netdisco_linux470->        usename,
netdisco_linux470->        pg_blocking_pids(pid) as blocked_by,
netdisco_linux470->        query as blocked_query
netdisco_linux470-> from pg_stat_activity
netdisco_linux470-> where cardinality(pg_blocking_pids(pid)) > 0;
  pid   | usename  | blocked_by |                                                                                  blocked_query                                 
--------+----------+------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 460309 | netdisco | {430771}   | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 )
 461468 | netdisco | {430771}   | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 )
 488078 | netdisco | {430771}   | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 )
 430769 | netdisco | {430771}   | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 )
 430771 | netdisco | {415750}   | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 )
(5 rows)

netdisco_linux470=> \q
[1264283] 2024-07-31 14:52:53  info psql: finished at Wed Jul 31 16:52:53 2024
[1264283] 2024-07-31 14:52:53  info psql: status done: psql session closed.
netdisco@linux470:~$
```

**Comment by JeroenvIS** [comment]
Now including the rest of the pg_stat_activity columns:

```
  pid   | blocked_by | datid  |      datname      |  pid   | leader_pid | usesysid | usename  | application_name | client_addr |     client_hostname      | client_port |         backend_start         |          xact_start           |          query_start          |         state_change          | wait_event_type |  wait_event   | state  | backend_xid | backend_xmin | query_id |                                                                                      query                                                                                       |  backend_type
--------+------------+--------+-------------------+--------+------------+----------+----------+------------------+-------------+--------------------------+-------------+-------------------------------+-------------------------------+-------------------------------+-------------------------------+-----------------+---------------+--------+-------------+--------------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------
 460309 | {430771}   | 311096 | netdisco_linux470 | 460309 |            |   309777 | netdisco | netdisco         | 10.20.0.12  | linux811.utsp.utwente.nl |       53758 | 2024-07-30 05:07:33.811574+02 | 2024-07-30 06:26:59.32098+02  | 2024-07-30 06:26:59.342766+02 | 2024-07-30 06:26:59.342767+02 | Lock            | transactionid | active |  2426222911 |   2425897188 |          | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 ) | client backend
 461468 | {430771}   | 311096 | netdisco_linux470 | 461468 |            |   309777 | netdisco | netdisco         | 10.20.0.12  | linux811.utsp.utwente.nl |       34992 | 2024-07-30 05:11:18.439183+02 | 2024-07-30 06:14:16.907332+02 | 2024-07-30 06:14:16.930588+02 | 2024-07-30 06:14:16.93059+02  | Lock            | transactionid | active |  2426219421 |   2425897188 |          | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 ) | client backend
 488078 | {430771}   | 311096 | netdisco_linux470 | 488078 |            |   309777 | netdisco | netdisco         | 10.20.0.12  | linux811.utsp.utwente.nl |       45252 | 2024-07-30 07:03:01.696908+02 | 2024-07-30 09:04:49.551109+02 | 2024-07-30 09:04:49.572522+02 | 2024-07-30 09:04:49.572523+02 | Lock            | transactionid | active |  2426283280 |   2425897188 |          | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 ) | client backend
 430769 | {430771}   | 311096 | netdisco_linux470 | 430769 |            |   309777 | netdisco | netdisco         | 10.20.0.12  | linux811.utsp.utwente.nl |       33050 | 2024-07-30 03:05:21.850676+02 | 2024-07-30 04:35:54.773253+02 | 2024-07-30 04:35:54.801722+02 | 2024-07-30 04:35:54.801722+02 | Lock            | transactionid | active |  2426104349 |   2425897188 |          | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 ) | client backend
 430771 | {415750}   | 311096 | netdisco_linux470 | 430771 |            |   309777 | netdisco | netdisco         | 10.20.0.12  | linux811.utsp.utwente.nl |       33072 | 2024-07-30 03:05:21.884767+02 | 2024-07-30 03:14:46.720289+02 | 2024-07-30 03:14:47.346874+02 | 2024-07-30 03:14:47.346874+02 | Lock            | transactionid | active |  2425993603 |   2425897188 |          | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 ) | client backend
(5 rows)
```
and the initial blocking query appears to be:
```
netdisco_linux470=> select * from pg_stat_activity where pid=415750;
 datid  |      datname      |  pid   | leader_pid | usesysid | usename  | application_name | client_addr |     client_hostname      | client_port |         backend_start         |          xact_start           |          query_start          |         state_change          | wait_event_type | wait_event |        state        | backend_xid | backend_xmin | query_id |                                                                                      query                                                                                       |  backend_type
--------+-------------------+--------+------------+----------+----------+------------------+-------------+--------------------------+-------------+-------------------------------+-------------------------------+-------------------------------+-------------------------------+-----------------+------------+---------------------+-------------+--------------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------
 311096 | netdisco_linux470 | 415750 |            |   309777 | netdisco | netdisco         | 10.20.0.12  | linux811.utsp.utwente.nl |       35270 | 2024-07-30 02:05:13.781717+02 | 2024-07-30 02:17:45.248816+02 | 2024-07-30 02:17:45.500374+02 | 2024-07-30 02:17:45.500432+02 | Client          | ClientRead | idle in transaction |  2425897188 |              |          | INSERT INTO device_port_properties ( ifindex, ip, port, raw_speed, remote_dns, remote_is_discoverable, remote_is_phone, remote_is_wap) VALUES ( $1, $2, $3, $4, $5, $6, $7, $8 ) | client backend
(1 row)
```

Strange that all these pg_stat timestamps are yesterday (Jul 30), while the `ps -ef` output also has a process that started on the Jul 22 and one that started Jul 26...

**Comment by JeroenvIS** [comment]
We've had a few issues while I was away on vacation, not much details collected about those.

Last night both our production and test instance exhibited the same problem, somewhere after 7  pm. They run on different servers, connect through different pg_bouncer instances to different Postgres database servers.

Now running `DBIC_TRACE=1 ./perl5/bin/localenv ./bin/netdisco-backend-fg` on our test server (in screen with 64k lines scrollback buffer), hoping that that will give more detail when the issue appears again.

**Comment by JeroenvIS** [comment]
The DBIC_TRACE didn't give any real insights yet. Noticed though that the deployment.yml also has a setting for log level; just set that to "debug" on our test install and "info" on production.

Our test env has been upgraded to 2.078.000 with a couple of local changes to add more logging to MCE tasks and it's running the netdisco-backend-fg.

Hopefully this will provide more insight, now we just wait for a couple of days/weeks.

**Comment by JeroenvIS** [comment]
First crash in 5 days, production environment. All "nd2" workers died and only the master remained:

```
netdisco@linux786:~/logs$ ps -ef | grep nd2
netdisco 3842258  122301  0 Sep12 ?        00:16:13 [nd2: master] <defunct>
netdisco 4182655 2361986  0 14:10 pts/27   00:00:00 grep nd2
netdisco@linux786:~/logs$
```

The log ends with **EV: error in callback (ignoring): Can't call method "push_read" on an undefined value at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/DNS.pm line 1281.**

```
[3960986] 2024-09-17 11:50:13  info pol (127): wrapping up arpnip job(69091223) - status done at Tue Sep 17 13:50:13 2024
[3960988] 2024-09-17 11:50:13  info pol (104): wrapping up arpnip job(69091221) - status done at Tue Sep 17 13:50:13 2024
[3960979] 2024-09-17 11:50:13  info pol (71): wrapping up arpnip job(69091216) - status done at Tue Sep 17 13:50:13 2024
[4052111] 2024-09-17 11:50:13  info applying role Poller to worker 127
[4052113] 2024-09-17 11:50:13  info applying role Poller to worker 104
[4052115] 2024-09-17 11:50:13  info applying role Poller to worker 71
[3960994] 2024-09-17 11:50:13  info pol (47): wrapping up arpnip job(69091215) - status done at Tue Sep 17 13:50:13 2024
[4052116] 2024-09-17 11:50:13  info applying role Poller to worker 47
[3909913] 2024-09-17 11:50:14  info pol (13): wrapping up arpnip job(69091220) - status done at Tue Sep 17 13:50:14 2024
[4052237] 2024-09-17 11:50:15  info applying role Poller to worker 13
[3960900] 2024-09-17 11:50:15  info pol (123): wrapping up arpnip job(69091229) - status done at Tue Sep 17 13:50:15 2024
[3960913] 2024-09-17 11:50:15  info pol (73): wrapping up arpnip job(69091179) - status done at Tue Sep 17 13:50:15 2024
[4052322] 2024-09-17 11:50:15  info applying role Poller to worker 123
[4052324] 2024-09-17 11:50:15  info applying role Poller to worker 73
[3960963] 2024-09-17 11:50:15  info pol (119): wrapping up arpnip job(69091233) - status done at Tue Sep 17 13:50:15 2024
[3960965] 2024-09-17 11:50:15  info pol (7): wrapping up arpnip job(69091218) - status done at Tue Sep 17 13:50:15 2024
[4052346] 2024-09-17 11:50:15  info applying role Poller to worker 119
[3960984] 2024-09-17 11:50:15  info pol (35): wrapping up arpnip job(69091217) - status done at Tue Sep 17 13:50:15 2024
[4052350] 2024-09-17 11:50:15  info applying role Poller to worker 7
[4052356] 2024-09-17 11:50:15  info applying role Poller to worker 35
EV: error in callback (ignoring): Can't call method "push_read" on an undefined value at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/DNS.pm line 1281.
```

So, an error message that we've seen before in the log, but not enough info to determine where this is called from, nor what the root cause of the failure is.

**Comment by JeroenvIS** [comment]
If I read the code in AnyEvent::DNS correctly, this path would only be hit when retrying a DNS lookup over TCP, after a UDP request was successful but had TC bit set:
http://cvs.schmorp.de/AnyEvent/lib/AnyEvent/DNS.pm?revision=1.160&view=markup#l1255

**Comment by JeroenvIS** [comment]
Since last week, I've been running with a couple of local changes to AnyEvent::DNS - a cluck upon entering the code path where TC bit is set. I've seen it being hit several times without issues, but also just prior to the crash.

For reference, relevant bit of the logging:
```
[886244] 2024-09-20 10:50:36 debug  [130.89.0.203] check_mac - VRRP mac [00:00:5e:00:01:55] - skipping
[886244] 2024-09-20 10:50:36 debug  [130.89.0.203] check_mac - VRRP mac [00:00:5e:00:01:13] - skipping
[886244] 2024-09-20 10:50:36 debug  resolving 12559 ARP entries with max 50 outstanding requests
[886244] Truncated DNS reply; res is $VAR1 = {
          'ra' => 1,
          'qd' => [
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in'
                    ]
                  ],
          'op' => 'query',
          'tc' => 1,
          'aa' => '',
          'qr' => 1,
          'ns' => [],
          'cd' => '',
          'rc' => 'noerror',
          'rd' => 1,
          'id' => 19379,
          'ar' => [],
          'ad' => '',
          'an' => [
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in',
                      30,
                      'ptr.mail-ru-i-40.inconsistent.iprev.rir-data.org'
                    ],
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in',
                      30,
                      'ptr.zohomail-eu-i-40.inconsistent.iprev.rir-data.org'
                    ],
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in',
                      30,
                      'ptr.proton-me-i-40.inconsistent.iprev.rir-data.org'
                    ],
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in',
                      30,
                      'ptr.mozmail-com-i-40.inconsistent.iprev.rir-data.org'
                    ],
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in',
                      30,
                      'ptr.aol-com-i-40.inconsistent.iprev.rir-data.org'
                    ],
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in',
                      30,
                      'ptr.outlook-com-i-40.inconsistent.iprev.rir-data.org'
                    ],
                    [
                    [
                      '204.172.87.192.in-addr.arpa',
                      'ptr',
                      'in',
                      30,
                      'ptr.yahoo-com-i-40.inconsistent.iprev.rir-data.org'
                    ]
                  ],
          '__' => 'K���20417287192in-addrarpa�2ptrmail-ru-i-40inconsistentiprerir-dataorg�6ptrzohomail-eu-i-40inconsistentiprerir-dataorg�4ptrproton-me-i-40inconsistentiprerir-dataorg�6ptrmozmail-com-i-40inconsistentiprerir-dataorg�2ptraol-com-i-40inconsistentiprerir-dataorg�6ptroutlook-com-i-40inconsistentiprerir-dataorg�4ptryahoo-com-i-40inconsistentiprerir-dataorg'
        };
 
 at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/DNS.pm line 1257.
        AnyEvent::DNS::__ANON__(HASH(0x563867830ae8)) called at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/DNS.pm line 1198
        AnyEvent::DNS::_feed(AnyEvent::DNS=HASH(0x563866bf7b30), "K\x{b3}\x{83}\x{80}\x{0}\x{1}\x{0}\x{7}\x{0}\x{0}\x{0}\x{0}\x{3}204\x{3}172\x{2}87\x{3}192\x{7}in-addr\x{4}arpa\x{0}\x{0}\x{c}\x{0}\x{1}\x{c0}\x{c}\x{0}\x{c}\x{0}\x{1}\x{0}\x{0}\x{0}\x{1e}\x{0}2\x{3}ptr"...) called at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/DNS.pm line 1211
        AnyEvent::DNS::_recv(AnyEvent::DNS=HASH(0x563866bf7b30), "K\x{b3}\x{83}\x{80}\x{0}\x{1}\x{0}\x{7}\x{0}\x{0}\x{0}\x{0}\x{3}204\x{3}172\x{2}87\x{3}192\x{7}in-addr\x{4}arpa\x{0}\x{0}\x{c}\x{0}\x{1}\x{c0}\x{c}\x{0}\x{c}\x{0}\x{1}\x{0}\x{0}\x{0}\x{1e}\x{0}2\x{3}ptr"..., "\x{2}\x{0}\x{0}5\x{82}Y\x{0}\x{80}\x{0}\x{0}\x{0}\x{0}\x{0}\x{0}\x{0}\x{0}") called at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/DNS.pm line 918
        AnyEvent::DNS::__ANON__(EV::IO=SCALAR(0x5638671fd558), 1) called at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/Impl/EV.pm line 88
        eval {...} called at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/Impl/EV.pm line 88
        AnyEvent::CondVar::Base::_wait(AnyEvent::CondVar=HASH(0x563866b13050)) called at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent.pm line 2031
        AnyEvent::CondVar::Base::recv(AnyEvent::CondVar=HASH(0x563866b13050)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Util/FastResolver.pm line 80
        App::Netdisco::Util::FastResolver::hostnames_resolve_async(ARRAY(0x56386785c138)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Worker/Plugin/Arpnip/Nodes.pm line 47
        App::Netdisco::Worker::Plugin::Arpnip::Nodes::__ANON__(App::Netdisco::Backend::Job=HASH(0x563864baa0b0), HASH(0x56386685bd10)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Worker/Plugin.pm line 117
        App::Netdisco::Worker::Plugin::__ANON__(App::Netdisco::Backend::Job=HASH(0x563864baa0b0)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Worker/Runner.pm line 105
        App::Netdisco::Worker::Runner::try {...} () called at /usr/share/perl5/Try/Tiny.pm line 102
        eval {...} called at /usr/share/perl5/Try/Tiny.pm line 93
        Try::Tiny::try(CODE(0x563866b069f8), Try::Tiny::Catch=REF(0x563867857fc8)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Worker/Runner.pm line 109
        App::Netdisco::Worker::Runner::run_workers(MCE__WITH__App::Netdisco::Backend::Role::Poller=HASH(0x563863f47ba0), "workers_store") called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Worker/Runner.pm line 74
        App::Netdisco::Worker::Runner::__ANON__(MCE__WITH__App::Netdisco::Backend::Role::Poller=HASH(0x563863f47ba0), App::Netdisco::Backend::Job=HASH(0x563864baa0b0)) called at /home/netdisco/perl5/lib/perl5/Sys/SigAction.pm line 164
        eval {...} called at /home/netdisco/perl5/lib/perl5/Sys/SigAction.pm line 162
        eval {...} called at /home/netdisco/perl5/lib/perl5/Sys/SigAction.pm line 160
        Sys::SigAction::timeout_call(600, CODE(0x563864b93e48), MCE__WITH__App::Netdisco::Backend::Role::Poller=HASH(0x563863f47ba0), App::Netdisco::Backend::Job=HASH(0x563864baa0b0)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Worker/Runner.pm line 82
        App::Netdisco::Worker::Runner::run(MCE__WITH__App::Netdisco::Backend::Role::Poller=HASH(0x563863f47ba0), App::Netdisco::Backend::Job=HASH(0x563864baa0b0)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Backend/Role/Poller.pm line 35
        App::Netdisco::Backend::Role::Poller::try {...} () called at /usr/share/perl5/Try/Tiny.pm line 102
        eval {...} called at /usr/share/perl5/Try/Tiny.pm line 93
        Try::Tiny::try(CODE(0x563864baef28), Try::Tiny::Catch=REF(0x563862832ea0)) called at /opt/netdisco/git/netdisco/lib/App/Netdisco/Backend/Role/Poller.pm line 41
        App::Netdisco::Backend::Role::Poller::worker_body(MCE__WITH__App::Netdisco::Backend::Role::Poller=HASH(0x563863f47ba0)) called at ./bin/netdisco-backend-fg line 81
        main::__ANON__(MCE__WITH__App::Netdisco::Backend::Role::Poller=HASH(0x563863f47ba0)) called at /home/netdisco/perl5/lib/perl5/MCE/Core/Worker.pm line 469
        MCE::_worker_do() called at /home/netdisco/perl5/lib/perl5/MCE/Core/Worker.pm line 687
        MCE::_worker_main() called at /home/netdisco/perl5/lib/perl5/MCE.pm line 2038
        MCE::_dispatch() called at /home/netdisco/perl5/lib/perl5/MCE.pm line 2082
        MCE::_dispatch_child() called at /home/netdisco/perl5/lib/perl5/MCE.pm line 891
        MCE::restart_worker("MCE") called at ./bin/netdisco-backend-fg line 64
        main::__ANON__(MCE__WITH__App::Netdisco::Backend::Role::Poller=HASH(0x563863f47ba0), HASH(0x563864103b20), 0) called at /home/netdisco/perl5/lib/perl5/MCE/Core/Manager.pm line 277
        MCE::__ANON__() called at /home/netdisco/perl5/lib/perl5/MCE/Core/Manager.pm line 996
        MCE::_output_loop() called at /home/netdisco/perl5/lib/perl5/MCE.pm line 1186
        MCE::run() called at /home/netdisco/perl5/lib/perl5/MCE/Flow.pm line 426
        MCE::Flow::run(HASH(0x5638640deaa0), CODE(0x5638640debf0), CODE(0x5638640ded88), CODE(0x56386408f338)) called at ./bin/netdisco-backend-fg line 65
EV: error in callback (ignoring): Can't call method "push_read" on an undefined value at /home/netdisco/perl5/lib/perl5/x86_64-linux-gnu-thread-multi/AnyEvent/DNS.pm line 1283.
```

**Comment by JeroenvIS** [comment]
Been in touch with the author of AnyEvent::DNS through IRC. He has a theory on how this issue might occur and suggested a local change in DNS.pm.
We've had several nd2 crashes over the last days, all with the same `Can't call method "push_read" on an undefined value`, so if the crashes don't occur with this change for at least a couple of weeks, then it's a good sign.

Agreed with AnyEvent author to try this and report back to him.

**Comment by JeroenvIS** [comment]
Looking back through the ticket, we've had two distinct issues:
  1. Several occasions where all nd2 processes are gone and a defunct "nd2: master" is the only one remaining,
  2. At least one occasion where all the workers are still active, but appear to be in a deadlock for database access.
 
The change in AnyEvent::DNS is likely to fix the first of these, which is the most frequent anyway.

**Comment by ollyg** [comment]
Well done @JeroenvIS ! amazing work and I'm glad the AnyEvent::DNS maintainer is supportive and reactive 💟 

I am following along, even if not able to work on it myself much right now.

**Comment by JeroenvIS** [comment]
Hi @ollyg, sorry, completely missed your update - thanks for the highlight on IRC.

TBH I've sort-of put this on hold. We don't have the issue that often any more and we're monitoring the mean ages of last_discover, last_macsuck and last_arpnip from the database as well as the amount of nd2 processes to be alerted when something breaks and needs manual intervention.

For a proper fix in the AnyEvent::DNS module, I'd like to have a reproducible test scenario first. I haven't written anything like that yet though. The cause was a UDP lookup receiving a reply with TC bit set, followed by a failure in the the retried lookup over TCP - iirc if the TCP connect for that retry fails, eg because of a RST. And I'm not sure about the best way to test that - purely simulated in code, or with eg netem simulating connect timeouts or injecting RSTs, etc.
